packages/contracts-bedrock/test/invariants/symbolic/OptimismSuperchainERC20.t.sol

```diff
@@ -0,0 +1,234 @@
+    string internal name = "SuperchainERC20";
+    string internal symbol = "SUPER";
+    uint8 internal decimals = 18;
+    address internal user = address(bytes20(keccak256("user")));
+    address internal target = address(bytes20(keccak256("target")));
+
+    OptimismSuperchainERC20 public superchainERC20Impl;
+    OptimismSuperchainERC20 internal optimismSuperchainERC20;
+
+    constructor() {
+        superchainERC20Impl = new OptimismSuperchainERC20();
+        optimismSuperchainERC20 = OptimismSuperchainERC20(
+            address(
+                new ERC1967Proxy(
+                    address(superchainERC20Impl),
+                    abi.encodeCall(OptimismSuperchainERC20.initialize, (remoteToken, name, symbol, decimals))
+                )
+            )
+        );
+
+        // Etch the mocked L2 to L2 Messenger because the `TSTORE` opcode is not supported, and also due to issues with
+        // `encodeVersionedNonce()`
+        address _mockL2ToL2CrossDomainMessenger = address(new MockL2ToL2Messenger(address(optimismSuperchainERC20)));
+        hevm.etch(address(MESSENGER), _mockL2ToL2CrossDomainMessenger.code);
+    }
+
+    function eqStrings(string memory a, string memory b) internal pure returns (bool) {
+        return keccak256(abi.encode(a)) == keccak256(abi.encode(b));
+    }
+
+    /// @custom:property-id 0
+    /// @custom:property Check setup works as expected
+    function check_setup() public view {
+        assert(optimismSuperchainERC20.remoteToken() == remoteToken);
+        assert(eqStrings(optimismSuperchainERC20.name(), name));
+        assert(eqStrings(optimismSuperchainERC20.symbol(), symbol));
+        assert(optimismSuperchainERC20.decimals() == decimals);
+    }
+
+    /// @custom:property-id 6
+    /// @custom:property-id Calls to sendERC20 succeed as long as caller has enough balance
+    function check_sendERC20SucceedsOnlyIfEnoughBalance(
+        uint256 _initialBalance,
+        uint256 _amount,
+        address _to,
+        uint256 _chainId
+    )
+        public
+    {
+        /* Preconditions */
+        vm.assume(_chainId != CURRENT_CHAIN_ID);
+        vm.assume(_to != address(0));
+
+        // Can't use symbolic value for user since it fails due to `NotConcreteError`
+        // hevm.deal(address(optimismSuperchainERC20), user, _initialBalance);
+        vm.prank(Predeploys.L2_STANDARD_BRIDGE);
+        optimismSuperchainERC20.mint(user, _initialBalance);
+
+        vm.prank(user);
+        /* Action */
+        try optimismSuperchainERC20.sendERC20(_to, _amount, _chainId) {
+            /* Postcondition */
+            assert(_initialBalance >= _amount);
+        } catch {
+            assert(_initialBalance < _amount);
+        }
+    }
+
+    /// @custom:property-id 7
+    /// @custom:property-id Calls to relayERC20 always succeed as long as the cross-domain caller is valid
+    function check_relayERC20OnlyFromL2ToL2Messenger(address _sender, uint256 _amount) public {
```

Comment:
on this method you're checking the call reverts if the caller is not the `L2ToL2CrossDomainMessenger`, however, the idea of this test was to see if we can get rid of the `msg.sender == MESSENGER` check because of the `MESSENGER.crossDomainMessagesender() != address(this)` already covering it:

        if (msg.sender != MESSENGER) revert CallerNotL2ToL2CrossDomainMessenger();
        if (IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSender() != address(this)) {
            revert InvalidCrossDomainSender();
        }

see relevant discussion: https://github.com/ethereum-optimism/optimism/pull/11256#discussion_r1706092088

do you think it's possible to prove it's equivalent to perform both checks vs the second check only?

I'm afraid of doing so would require to bringing into scope a few more contracts (which use transient storage) so idk what to do about it

> @notice Partially verified since it can't be fully verified due to the use of `crossDomainMessageSender()`

perhaps we could prove it while mocking its return value? like we currently do with the sender

Reply:
Yeah, I tried that but I couldn't find a way to make it work.
When executing the setter on `forTest_setCrossDomainCaller` it always fails and idk why yet